### PR TITLE
14110 activate bundles on install

### LINKED ISSUE #14110: [Improvement][Installer] Ask for activation of in repository included bundles during install
### Improvement description

... so that developer can decide, which bundles should be activated directly during installation, e.g. SimpleBackendSearch

related to https://github.com/pimcore/pimcore/issues/12094

Only relevant for directly in pimcore/pimcore repository located bundles. Classic Admin UI probably will be an exception.

### PR DESCRIPTION
**Pull request by mattamon**
## Changes in this pull request  
Resolves #14110 

## Additional info  
Activating bundles via installer

### CODE REVIEW

**Comment by jdreesen** [comment]
This only runs on the first installation, right? Because I wouldn't like it if Pimcore messed up my `bundles.php` later on.

**Comment by mattamon** [comment]
@jdreesen This can run multiple times. The default value for the question if you want to install bundles is `no`. 
We should add an option to skip this question overall though. 

https://github.com/pimcore/pimcore/pull/14194/files#diff-192cfd82163e1c9667c58ac33150baa2511d7eaf23fcd2a787d1c0586426041bR26 
We do include the bundles.php and merge only the new bundles into the existing bundles.php

Do you have any other concerns?

**Comment by jdreesen** [comment]
But it'll change the style, right?

I hope it's at least like:
```php
<?php

return [
    Symfony\Bundle\FrameworkBundle\FrameworkBundle::class => ['all' => true],
    Symfony\Bundle\TwigBundle\TwigBundle::class => ['all' => true],
];
```
And not like:
```php
<?php

return [
    'Symfony\\Bundle\\FrameworkBundle\\FrameworkBundle' => ['all' => true],
    'Symfony\\Bundle\\TwigBundle\\TwigBundle' => ['all' => true]
];
```
But I'd prefer if it would keep the style if it's a custom one, so things like https://github.com/pimcore/pimcore/blob/e362a81192a5aba3a65fe03de1fd62d16b891804/.github/ci/files/config/bundles.php or https://github.com/Orchesty/orchesty-tutorial/blob/631fe4ad1c20c011b2128c3e35e87b922f627862/php-sdk/config/Bundles.php stay the way they are.

**Comment by mattamon** [comment]
Unfortunately, yes it will change the style.
By automatically generating the bundles.php the first style is not possible, unless you can provide a good way how to deal with this but as far as I know these constants will always be resolved unless written manually this way.
```php
<?php

return [
    Symfony\Bundle\FrameworkBundle\FrameworkBundle::class => ['all' => true],
    Symfony\Bundle\TwigBundle\TwigBundle::class => ['all' => true],
];
```
Those were my initial concerns too. 

On the other hand, why would you call the installer on a running and set up instance? 
This should not be the primary way on how to install bundles, this should just make it easier for the first time installation.

**Review by dvesh3 — COMMENTED** [review]
@mattamon How about we also add an option to install the bundles in no interaction mode, as this will help in [automated installations](https://pimcore.com/docs/pimcore/current/Development_Documentation/Getting_Started/Advanced_Installation_Topics.html#page_Advanced-Installation-Topics). We can pass a CLI parameter like 
```
/vendor/bin/pimcore-install \
  --admin-username=admin --admin-password=admin \
  --mysql-database=pimcore \
  --no-interaction
  --install-bundles=PimcoreGlossaryBundle,PimcoreSeoBundle,PimcoreUuidBundle
```
wdyt?

**Comment by jdreesen** [comment]
> By automatically generating the `bundles.php` the first style is not possible, unless you can provide a good way how to deal with this but as far as I know these constants will always be resolved unless written manually this way.

Well, it should be possible to define the keys like `"{$bundle}::class"`, right?

> On the other hand, why would you call the installer on a running and set up instance?
This should not be the primary way on how to install bundles, this should just make it easier for the first time installation.

Right, therefore my question, if the file is only adjusted during the first installation, or if you plan to also do this later on (like the previous `bin/console pimcore:bundle:enable FooBundle`)

**Comment by mattamon** [comment]
> Well, it should be possible to define the keys like `"{$bundle}::class"`, right?

What about use statements? If someone is using a use statement they will be removed. So it would not look the same for them and for everyone who is not using the ::class it will change for them too.

> Right, therefore my question, if the file is only adjusted during the first installation, or if you plan to also do this later on (like the previous `bin/console pimcore:bundle:enable FooBundle`)

At the current state the bundles.php will always change if you decide to install a bundle via the pimcore install command.

**Comment by dpfaffenbauer** [comment]
symfony flex is also able to change the bundles.php without removing use statements or changing the style, maybe you look into how they do it?

**Comment by jdreesen** [comment]
I didn't use flex myself yet, but this should be the script, right? 
https://github.com/symfony/flex/blob/2.x/src/Configurator/BundlesConfigurator.php

In the first sight it doesn't look like it keeps the style, though.

**Comment by mattamon** [comment]
@dvesh3 added the option install-bundles. You can check all the bundles in the description using --help

@jdreesen I copied the function `buildContents` from `symfony/flex` 
The bundles.php will look like this now
```php
return [
    Pimcore\Bundle\SeoBundle\PimcoreSeoBundle::class => ['all' => true],
];
```

**Comment by dpfaffenbauer** [comment]
> I didn't use flex myself yet, but this should be the script, right? 
> 
> https://github.com/symfony/flex/blob/2.x/src/Configurator/BundlesConfigurator.php
> 
> 
> 
> In the first sight it doesn't look like it keeps the style, though.

Ok, doesn't keep it, but at least it builds it right ;)

File: bundles/InstallBundle/src/BundleConfig/BundleWriter.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+use Symfony\Component\Filesystem\Exception\FileNotFoundException;
+
+class BundleWriter
+{
+    public function addBundlesToConfig(array $bundles): void
+    {
+        $bundlesPhpFile = PIMCORE_PROJECT_ROOT . '/config/bundles.php';
+
+        if (!file_exists($bundlesPhpFile)) {
+            throw new FileNotFoundException();
```
Comment (dvesh3) [inline bundles/InstallBundle/src/BundleConfig/BundleWriter.php]:
```suggestion
            throw new FileNotFoundException("File \"$bundlesPhpFile\" not found!");
```

File: bundles/InstallBundle/src/Command/InstallCommand.php [hunk 2]
```diff
@@ -111,6 +111,11 @@ private function getOptions(): array
                 'default' => '',
                 'group' => 'db_credentials',
             ],
+            'install-bundles' => [
```
Comment (dvesh3) [inline bundles/InstallBundle/src/Command/InstallCommand.php:114]: Even when providing `--install-bundles` option, interactive prompt is still there to select bundles to install.

**Review by dvesh3 — COMMENTED** [review]
On re-install, using the option to install bundles clear out the content of bundles.php:
image

File: bundles/InstallBundle/src/BundleConfig/BundleWriter.php (hunk 1 above)
Comment (mattamon) [reply bundles/InstallBundle/src/BundleConfig/BundleWriter.php]: I could not reproduce the error that the bundles.php is cleared. 
Can you show me the command with the options you entered? 

I added the error message

Comment (dvesh3) [reply bundles/InstallBundle/src/BundleConfig/BundleWriter.php]: Seems to work fine now. I can't reproduce it anymore 😄

**Review by dvesh3 — APPROVED** [review]

**Pull request merged**